cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultEntityEnsurer.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright © 2015-2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.namespace;
+
+import co.cask.cdap.common.AlreadyExistsException;
+import co.cask.cdap.common.NamespaceAlreadyExistsException;
+import co.cask.cdap.common.NotFoundException;
+import co.cask.cdap.common.namespace.NamespaceAdmin;
+import co.cask.cdap.common.service.RetryOnStartFailureService;
+import co.cask.cdap.common.service.RetryStrategies;
+import co.cask.cdap.internal.app.store.profile.ProfileStore;
+import co.cask.cdap.proto.NamespaceMeta;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProfileId;
+import co.cask.cdap.proto.profile.Profile;
+import com.google.common.util.concurrent.AbstractService;
+import com.google.common.util.concurrent.Service;
+import com.google.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.file.FileAlreadyExistsException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Thread that ensures that the default namespace exists
+ */
+public final class DefaultEntityEnsurer extends AbstractService {
+
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultEntityEnsurer.class);
+
+  private final Service serviceDelegate;
+
+  @Inject
+  public DefaultEntityEnsurer(NamespaceAdmin namespaceAdmin, ProfileStore profileStore) {
+    this.serviceDelegate = new RetryOnStartFailureService(() -> new AbstractService() {
+      @Override
+      protected void doStart() {
+        try {
+          if (!namespaceAdmin.exists(NamespaceId.DEFAULT)) {
+            namespaceAdmin.create(NamespaceMeta.DEFAULT);
+            // if there is no exception, assume successfully created and break
+            LOG.info("Successfully created namespace '{}'.", NamespaceMeta.DEFAULT);
+          }
+        } catch (FileAlreadyExistsException e) {
+          LOG.warn("Got exception while trying to create namespace '{}'.", NamespaceMeta.DEFAULT, e);
+          // avoid retrying if its a FileAlreadyExistsException
+        } catch (NamespaceAlreadyExistsException e) {
+          // default namespace already exists
+          LOG.info("Default namespace already exists.");
+        } catch (Exception e) {
```

Comment:
I think here if authorization is enabled, we will not be able to pass the auth check to create default namespace if no privileges are granted at cdap. Do we want to make sure the default namespace exists before we check if the default profile exists since the default profile should be considered in SYSTEM namespace right?

Reply:
i didn't make any changes to this part. do you know how it works today?

Reply:
oh I see what you're saying. You mean this loops forever and fails if auth is enabled? I can change it so that both of these operations are always run instead of the default profile getting skipped if default namespace creation fails.

Reply:
So the default namespace does not necessarily exist, if cdap does not have privilege, this ensurer will always fail, and it will retry until the default namespace is created(when user grants cdap the privilege). So I think we may need to have a separate check for the default profile.